sklearn/ensemble/_iforest.py

```diff
@@ -485,15 +508,28 @@ def _average_path_length(n_samples_leaf):
     n_samples_leaf = n_samples_leaf.reshape((1, -1))
     average_path_length = np.zeros(n_samples_leaf.shape)
 
-    mask_1 = n_samples_leaf <= 1
-    mask_2 = n_samples_leaf == 2
-    not_mask = ~np.logical_or(mask_1, mask_2)
+    mask_small = n_samples_leaf < 52
+    not_mask = ~mask_small
+
+    average_path_length[mask_small] = _average_path_length_small[
+        n_samples_leaf[mask_small]]
+
+    # Average path length equals 2*(H(n)-1), with H(n) the nth harmonic number.
+    # For the harmonic number calculation,
+    # see the following publications and references therein
+    # Villarino, M.B. Ramanujan’s Harmonic Number Expansion Into Negative
+    # Powers Of A Triangular Number. JIPAM. J. Inequal. Pure Appl. Math. 9(3),
+    # 89 (2008). https://www.emis.de/journals/JIPAM/article1026.html?sid=1026.
+    # Preprint at https://arxiv.org/abs/0707.3950.
+    # or
+    # Wang, W. Harmonic Number Expansions of the Ramanujan Type.
+    # Results Math 73, 161 (2018). https://doi.org/10.1007/s00025-018-0920-8
 
-    average_path_length[mask_1] = 0.
-    average_path_length[mask_2] = 1.
+    tmp = 1.0/np.square(n_samples_leaf[not_mask])
     average_path_length[not_mask] = (
-        2.0 * (np.log(n_samples_leaf[not_mask] - 1.0) + np.euler_gamma)
-        - 2.0 * (n_samples_leaf[not_mask] - 1.0) / n_samples_leaf[not_mask]
+        2.0 * (np.log(n_samples_leaf[not_mask]) - 1.0 + np.euler_gamma)
+        + 1.0/n_samples_leaf[not_mask]
+        - tmp*(1.0/6.0 - tmp*(1.0/60.0 - tmp/126.0))
```

Comment:
Can you give a better name to `tmp`? Also how can one find those constants here?

Reply:
Sure, renamed to `n2_inv`.

Reply:
About the second question, the coefficients are from the Euler asymptotic expansion of the harmonic numbers `H(n)` as found on Wikipedia (https://en.wikipedia.org/wiki/Harmonic_number#Calculation) or in the references provided above. The coefficients just have to be doubled since the path length is `2*(H(n)-1)`.